persist/fs/files.go

```diff
@@ -42,11 +42,105 @@ var timeZero time.Time
 
 const (
 	dataDirName       = "data"
+	snapshotDirName   = "snapshots"
 	commitLogsDirName = "commitlogs"
 )
 
 type fileOpener func(filePath string) (*os.File, error)
 
+// FilesetFile represents a set of Fileset files for a given block start
+type FilesetFile struct {
+	BlockStart time.Time
+	Files      []string
+}
+
+// FilesetFilesSlice is a slice of FilesetFile
+type FilesetFilesSlice []FilesetFile
+
+// Filepaths flattens a slice of FilesetFiles to a single slice of filepaths
+func (f FilesetFilesSlice) Filepaths() []string {
+	flattened := []string{}
+	for _, fileset := range f {
+		flattened = append(flattened, fileset.Files...)
+	}
+
+	return flattened
+}
+
+// NewFilesetFile creates a new Fileset file
+func NewFilesetFile(blockStart time.Time) FilesetFile {
+	return FilesetFile{
+		BlockStart: blockStart,
+		Files:      []string{},
+	}
+}
+
+// SnapshotFile represents a set of Snapshot files for a given block start
+type SnapshotFile struct {
+	// There can be multiple snapshot files with the same blockstart, so they
+	// also have an auto-incrementing index
+	Index int
+	FilesetFile
+}
+
+// HasCheckpointFile returns a bool indicating whether the given set of
+// snapshot files has a checkpoint file.
+func (s SnapshotFile) HasCheckpointFile() bool {
+	for _, fileName := range s.Files {
+		if strings.Contains(fileName, checkpointFileSuffix) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// SnapshotFilesSlice is a slice of SnapshotFile
+type SnapshotFilesSlice []SnapshotFile
+
+// Flatten flattens a slice of SnapshotFiles to a single slice of filepaths
+func (f SnapshotFilesSlice) Flatten() []string {
+	flattened := []string{}
+	for _, snapshot := range f {
+		flattened = append(flattened, snapshot.Files...)
+	}
+
+	return flattened
+}
+
+// LatestForBlock returns the latest (highest index) SnapshotFile in the
+// slice for a given block start.
+func (f SnapshotFilesSlice) LatestForBlock(blockStart time.Time) (SnapshotFile, bool) {
+	for i := 0; i < len(f); i++ {
+		curr := f[i]
+		if curr.BlockStart.Equal(blockStart) {
+			isEnd := (i == len(f)-1)
+			isHighestIdx := true
+			if !isEnd {
+				next := f[i+1]
+				if next.BlockStart.Equal(blockStart) && next.Index > curr.Index {
```

Comment:
This relies on a sort order yeah?  Or are you relying on there only being two snapshots for the same time?  Is it worth just doing a second for loop here and ruling out it's definitely not the highest?  Such tiny number of files would make the time complexity basically just the same on average.

Reply:
yeah they're always returned sorted by time and index ascending

Do you mean just pretend they're not sorted and do two nested for loops?

Reply:
solved by forcing a sort at the top